### PR TITLE
digraph: Implementation of a reflexive directed graph

### PR DESCRIPTION
**Pull request by Delja**
Add the implementation of a reflexive directed graph. Added element is in relation with itself (ie if the graph has `u` node is implies `self.has_arc(u,u) == true`)

### CODE REVIEW

File: lib/graph/digraph.nit [hunk 1]
```diff
@@ -946,3 +946,133 @@ class HashDigraph[V: Object]
 
 	redef fun vertices_iterator: Iterator[V] do return outgoing_vertices_map.keys.iterator
 end
+
+# A reflexive directed graph
```
Comment (Morriar) [inline lib/graph/digraph.nit]: Can you add more documentation about why this is useful and why it cannot be achieved with `HashDigraph` please?

Comment (Delja) [reply lib/graph/digraph.nit]: `HashDigraph` can manage the reflexive links but these must be explicitly indicated by the user. Creating another specific class for that (`ReflexiveHashDigraph`) avoids having to define the links manually as reflexive.

File: lib/graph/digraph.nit [hunk 2]
```diff
@@ -946,3 +946,133 @@ class HashDigraph[V: Object]
...
+
+	# Adds the arc (u,v) to this graph.
+	# if `u` is the same as `v` do nothing
+	#
+	# ~~~
+	# var g = new ReflexiveHashDigraph[Int]
+	# g.add_arc(1, 2)
+	# g.add_arc(2, 2)
+	# g.add_arc(3, 1)
+	# assert g.has_arc(2,2)
```
Comment (Morriar) [inline lib/graph/digraph.nit]: You should also give an assert about `(1, 2)` or `(3,1)`

File: lib/graph/digraph.nit [hunk 3]
```diff
@@ -946,3 +946,133 @@ class HashDigraph[V: Object]
...
+	# ~~~
+	redef fun add_arc(u, v)
+	do
+		# Check `u` is the same as `v`
+		if u != v then
+			super
+		end
+	end
+
+	# Is (u,v) an arc in this graph ?
```
Comment (Morriar) [inline lib/graph/digraph.nit]: There is no space before a question mark in english.

**Comment by Delja** [comment]
```
var reflexive_digraph = new ReflexiveHashDigraph[Int]
reflexive_digraph.add_vertex(1)
assert reflexive_digraph.has_arc(1,1)

var digraph = new HashDigraph[Int]
digraph.add_vertex(1)
digraph.add_arc(1,1) # The line is needed if your want to `1` has_arc on itself
assert digraph.has_arc(1,1)
```
Here is a comparison between the two classes of graphs. Both have the same goal, the reason of  `ReflexiveHashDigraph` it's just to remove the `add_arc` step.

@Morriar The motivation is simply to remove a potential mistake when you need all vertices to be reflexive. The other way would be to add a method in HashDigraph to make all the vertices reflexive, but for me it's a better solution to add a specific class to do that.

**Pull request merged**